### PR TITLE
feat(udf): Fixing cloning behavior, adding unit tests for custom user data.

### PR DESCRIPTION
**Pull request by dbakshiyev**

### CODE REVIEW

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy [hunk 1]
```diff
@@ -150,12 +153,26 @@ class CopyLastGoogleServerGroupAtomicOperation extends GoogleAtomicOperation<Dep
...
+        if (description.instanceMetadata) {
+          // Keep previous custom user data, unless new user data is also specified directly on the cloneServerGroup call.
+          if (description.userData) {
+            newDescription.instanceMetadata = description.instanceMetadata
+            newDescription.userData = description.userData
+          } else {
+            // If the user doesn't specify new custom user data, we want to copy the old value.
+            for (item in instanceMetadata.getItems()) {
+              // Look for custom user data.
+              if (item.key == 'customUserData') {
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy]: If the group being cloned had customer user data specified, won't all of the attributes (including `customUserData`) already be defined as key/value pairs on the instance metadata?

**Comment by duftler** [comment]
Unit tests seem to have failed.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy (hunk 1 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy]: Discussed offline. I get it now.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy [hunk 2]
```diff
@@ -150,12 +153,26 @@ class CopyLastGoogleServerGroupAtomicOperation extends GoogleAtomicOperation<Dep
...
-            ? description.instanceMetadata
-            : GCEUtil.buildMapFromMetadata(instanceMetadata)
+        if (description.instanceMetadata) {
+          // Keep previous custom user data, unless new user data is also specified directly on the cloneServerGroup call.
+          if (description.userData) {
+            newDescription.instanceMetadata = description.instanceMetadata
+            newDescription.userData = description.userData
+          } else {
+            // If the user doesn't specify new custom user data, we want to copy the old value.
+            for (item in instanceMetadata.getItems()) {
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy]: This being groovy, can do something like:

```
def item = instanceMetada.getItems().find { it.key == 'customUserData }

if (item) {
...
}
```

Also, can `getItems()` be `null`? If so, should use null-safe navigation:

`def item = instanceMetada.getItems()?.find { it.key == 'customUserData}`

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 3]
```diff
@@ -258,7 +260,120 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
 
     then:
       1 * googleClusterProviderMock.getServerGroup(ACCOUNT_NAME, REGION, ANCESTOR_SERVER_GROUP_NAME) >> serverGroup
-
       1 * basicGoogleDeployHandlerMock.handle(newDescription, _) >> deploymentResult
   }
+
+  void "operation builds description based on ancestor server group; overrides instanceMetadata and userData"() {
+    setup:
+    def newInstanceMetadata = ["differentKey": "differentValue"]
```
Comment (dbakshiyev) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Done

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 4]
```diff
@@ -258,7 +260,120 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
...
+    def newCustomUserdata = ["diff val": "different data"]
+    def description = new BasicGoogleDeployDescription(source: [region: REGION,
+                                                                serverGroupName: ANCESTOR_SERVER_GROUP_NAME],
+      region: REGION,
```
Comment (dbakshiyev) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Done

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 5]
```diff
@@ -258,7 +260,120 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
...
+    operation.registry = registry
+    operation.googleClusterProvider = googleClusterProviderMock
+    operation.basicGoogleDeployHandler = basicGoogleDeployHandlerMock
+
+    when:
+    operation.operate([])
+
+    then:
+    1 * googleClusterProviderMock.getServerGroup(ACCOUNT_NAME, REGION, ANCESTOR_SERVER_GROUP_NAME) >> serverGroup
+    //1 * basicGoogleDeployHandlerMock.handle({ println "$it"; return true}, _) >> deploymentResult
```
Comment (dbakshiyev) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Done

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 6]
```diff
@@ -258,7 +260,120 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
...
+    1 * basicGoogleDeployHandlerMock.handle(newDescription, _) >> deploymentResult
+  }
+  void "operation builds description based on ancestor server group; overrides userData"() {
```
Comment (dbakshiyev) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Done

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy (hunk 2 above)
Comment (dbakshiyev) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy]: Done

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 7]
```diff
@@ -258,7 +259,118 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
...
+      newDescription.onHostMaintenance = BaseGoogleInstanceDescription.OnHostMaintenance.MIGRATE
+      newDescription.serviceAccountEmail = SERVICE_ACCOUNT_EMAIL
+      newDescription.authScopes = DECORATED_AUTH_SCOPES
+      newDescription.network = DEFAULT_NETWORK_NAME
+      newDescription.subnet = SUBNET_NAME
+      newDescription.associatePublicIpAddress = true
+      newDescription.canIpForward = false
+      newDescription.loadBalancers = LOAD_BALANCERS
+      newDescription.autoscalingPolicy = new GoogleAutoscalingPolicy(coolDownPeriodSec: 45,
+        minNumReplicas: 2,
```
Comment (duftler) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Alignment is off here.

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 8]
```diff
@@ -258,7 +259,118 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
...
+        maxNumReplicas: 5)
+      newDescription.autoHealingPolicy = new GoogleAutoHealingPolicy(healthCheck: 'some-health-check',
+        initialDelaySec: 600)
```
Comment (duftler) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Alignment is off here.

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 9]
```diff
@@ -258,7 +259,118 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
...
+      newDescription.onHostMaintenance = BaseGoogleInstanceDescription.OnHostMaintenance.MIGRATE
+      newDescription.serviceAccountEmail = SERVICE_ACCOUNT_EMAIL
+      newDescription.authScopes = DECORATED_AUTH_SCOPES
+      newDescription.network = DEFAULT_NETWORK_NAME
+      newDescription.subnet = SUBNET_NAME
+      newDescription.associatePublicIpAddress = true
+      newDescription.canIpForward = false
+      newDescription.loadBalancers = LOAD_BALANCERS
+      newDescription.autoscalingPolicy = new GoogleAutoscalingPolicy(coolDownPeriodSec: 45,
+        minNumReplicas: 2,
```
Comment (duftler) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Alignment is off here.

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy [hunk 10]
```diff
@@ -258,7 +259,118 @@ class CopyLastGoogleServerGroupAtomicOperationUnitSpec extends Specification {
...
+        maxNumReplicas: 5)
+      newDescription.autoHealingPolicy = new GoogleAutoHealingPolicy(healthCheck: 'some-health-check',
+        initialDelaySec: 600)
```
Comment (duftler) [inline clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperationUnitSpec.groovy]: Alignment is off here.

**Pull request merged**